pkg/mic/mic.go

```diff
@@ -205,10 +245,64 @@ func (c *Client) NewLeaderElector(clientSet *kubernetes.Clientset, recorder reco
 	return leaderElector, nil
 }
 
+func (c *Client) UpgradeTypeIfRequired() error {
+	if c.TypeUpgradeCfg.EnableTypeUpgrade {
+		cm, err := c.CMClient.Get(c.CMCfg.Name, v1.GetOptions{})
+		// If we get an error and its not NotFound then return, because we cannot proceed.
+		if err != nil && !kubeErrors.IsNotFound(err) {
+			return fmt.Errorf("config map get for %s failed with error: %v", c.CMCfg.Name, err)
+		}
+
+		// Now either the configmap is not there or we successfully got the configmap
+		// Handle the case where the configmap is not found.
+		if err != nil && kubeErrors.IsNotFound(err) {
+			// Create the configmap
+			newCfgMap := &corev1.ConfigMap{
+				ObjectMeta: v1.ObjectMeta{
+					Namespace: c.CMCfg.Namespace,
+					Name:      c.CMCfg.Name,
+				},
+			}
+			if cm, err = c.CMClient.Create(newCfgMap); err != nil {
+				return fmt.Errorf("create configmap %s/%s failed with error: %v", c.CMCfg.Namespace, c.CMCfg.Name, err)
+			}
+		}
+
+		// We reach here only if the configmap is present or we created new one.
+		// Check if the key for type upgrade is present. If the key is present,
+		// then the upgrade is already performed. If not then go through the type upgrade
+		// process.
+		if v, ok := cm.Data[c.TypeUpgradeCfg.CMTypeUpgradeKey]; !ok {
+			klog.Infof("Upgrading the types to work with case sensitive go-client")
+			if err := c.CRDClient.UpgradeAll(); err != nil {
+				return fmt.Errorf("type upgrade failed. error: %+v", err)
+			}
+			klog.Infof("Type upgrade completed !!")
+			// Upgrade completed so update the data with the upgrade key.
+			if cm.Data == nil {
+				cm.Data = make(map[string]string)
+			}
+			cm.Data[c.TypeUpgradeCfg.CMTypeUpgradeKey] = version.MICVersion
+			_, err = c.CMClient.Update(cm)
+			if err != nil {
+				return fmt.Errorf("type upgrade annotation update on %s failed. error: %+v", c.TypeUpgradeCfg.CMTypeUpgradeKey, err)
+			}
+		} else {
+			klog.Infof("Type upgrade status configmap found from version: %s. Proceeding without type upgrade !", v)
+		}
+	}
+	return nil
+}
+
 // Start ...
 func (c *Client) Start(exit <-chan struct{}) {
 	klog.V(6).Infof("MIC client starting..")
 
+	if err := c.UpgradeTypeIfRequired(); err != nil {
+		klog.Fatalf("Upgrade failed with error: %v", err)
```

Comment:
I think we should say type upgrade failed with error. I think we need to make this error message descriptive so the user what upgrade failed here